crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -131,41 +131,39 @@ where
             return Ok(headers.into_iter().cloned().map(BlockResponse::Empty).collect())
         }
 
-        let request_len = headers_with_txs_and_ommers.len();
-        tracing::trace!(target: "downloaders::bodies", request_len, "Requesting bodies");
-        let (peer_id, bodies) =
-            self.client.get_block_bodies(headers_with_txs_and_ommers).await?.split();
-        tracing::trace!(
-            target: "downloaders::bodies", request_len, response_len = bodies.len(), ?peer_id, "Received bodies"
-        );
+        let mut responses = Vec::with_capacity(headers_with_txs_and_ommers.len());
+        while responses.len() != headers_with_txs_and_ommers.len() {
+            let request: Vec<_> =
+                headers_with_txs_and_ommers.iter().skip(responses.len()).cloned().collect();
+            let request_len = request.len();
+            tracing::trace!(target: "downloaders::bodies", request_len, "Requesting bodies");
+            let (peer_id, bodies) = self.client.get_block_bodies(request.clone()).await?.split();
 
-        let mut bodies = bodies.into_iter();
+            if bodies.is_empty() {
+                tracing::error!(
+                    target: "downloaders::bodies", ?peer_id, ?request, "Received empty response. Penalizing peer"
+                );
+                self.client.report_bad_message(peer_id);
+                // return error and trigger new retry
+                return Err(DownloadError::RequestError(RequestError::BadResponse))
+            }
```

Comment:
one thing to note is that this case discards all of the downloaded bodies. maybe it's worth having an internal retry here as well